File: Codility/11_sieve_of_eratosthenes/count_semiprimes.cpp
vector<int> solution(int N, vector<int> &P, vector<int> &Q) {
    vector<int> counters(N + 1);
    for (long long i = 2; i <= N; ++i)
    {
        for (long long j = 2 * i; j <= N; j += i)
            ++counters[j];
            
        for (long long j = i * i, k = 1; j <= N; j *= i, ++k)
            counters[j] += k;
    }

    vector<int> semiprimes_less{0};
        
    for (int i = 1; i <= N; ++i)
    {
        if (counters[i] == 2)
            semiprimes_less.push_back(semiprimes_less[i - 1] + 1);
        else
            semiprimes_less.push_back(semiprimes_less[i - 1]);
    }

    vector<int> result;
    for (int i = 0; i < P.size(); ++i)
    {
        result.push_back(semiprimes_less[Q[i]] - semiprimes_less[P[i] - 1]);
    }
    
    return result;
}

File: Codility/11_sieve_of_eratosthenes/count_semiprimes.rb
# https://codility.com/programmers/task/count_semiprimes/

def solution(n, p, q)
  factor_counts = Array.new(n + 1, 0) 
  for i in 2..n
    (2 * i..n).step(i) { |j| factor_counts[j] += 1 }
    
    j = i * i
    while j <= n
      factor_counts[j] += 1
      j *= i
    end
  end
  
  semiprimes_count_less = [ 0 ]
  for factor_count in factor_counts[1..-1]
    if factor_count == 2
      semiprimes_count_less << semiprimes_count_less.last + 1
    else
      semiprimes_count_less << semiprimes_count_less.last
    end
  end

  result = []
  p.zip(q).each do |pe, qe|
    result << semiprimes_count_less[qe] - semiprimes_count_less[pe - 1]
  end
  
  result
end
